### PR TITLE
add volumeName for each mountpoint

### PR DESCRIPTION
**Pull request by vohtaski**
Introduced a name `volumeName` next to path in `mountpoints` array. Addresses #123 for windows and mac. Linux is needed.

### CODE REVIEW

File: scripts/darwin.sh [hunk 1]
```diff
@@ -61,6 +61,7 @@ for disk in $DISKS; do
     echo "mountpoints:"
     echo "$mountpoints" | while read -r mountpoint ; do
       echo "  - path: \"$mountpoint\""
+      echo "    volumeName: \"$mountpoint\""
```
Comment (lurch) [inline scripts/darwin.sh]: Huh? You're setting both `path` and `volumeName` to the same `$mountpoint` variable??

File: scripts/win32.bat [hunk 2]
```diff
@@ -128,12 +129,14 @@ Function GetTopLevelDrives()
 			Summary.Add "Size", TopLevelDrive.Size
 
 			Set Mountpoints = new List
+			Set MountpointObjects = new List
```
Comment (lurch) [inline scripts/win32.bat]: Could the code be tidied up to use _just_ `MountpointObjects`, and then get rid of `Mountpoints` ? (to avoid the duplication). I'm afraid I don't know VBScript very well myself.

File: scripts/win32.bat [hunk 3]
```diff
@@ -145,7 +148,7 @@ Function GetTopLevelDrives()
 				End If
 			Next
 
-			Summary.Add "Mountpoints", Mountpoints
+			Summary.Add "Mountpoints", MountpointObjects
```
Comment (lurch) [inline scripts/win32.bat]: Perhaps the `"Mountpoints"` here (and on line 178) should be changed to `"MountpointObjects"` ?

File: lib/parse.js [hunk 4]
```diff
@@ -68,7 +68,7 @@ module.exports = (input) => {
     device = _.chain(device)
       .split('\n')
       .filter((line) => {
-        return /^(\s\s-\s)?[a-z]+:/g.test(line);
+        return /^(\s\s[-\s]\s)?[a-zA-Z]+:/g.test(line);
```
Comment (lurch) [inline lib/parse.js]: Could you explain why this change was needed?

**Comment by lurch** [comment]
And it would be great if this PR could update the examples in the README too :-)

**Comment by lurch** [comment]
@hedss What's happening with `Versionist` in this PR's status-checks?

File: scripts/darwin.sh (hunk 1 above)
Comment (vohtaski) [inline scripts/darwin.sh]: On mac, `path` already has the name of the device `/Volumes/Device_115`.
Not sure what's the best here, probably to parse the string and get `Device_115`.

File: lib/parse.js (hunk 4 above)
Comment (vohtaski) [inline lib/parse.js]: before parsing the output with yaml, this filter is applied, so it removes the line I added earlier.
"    volumeName: "

File: scripts/win32.bat (hunk 2 above)
Comment (vohtaski) [inline scripts/win32.bat]: Probably yes, but I also don't know much VBScript.
There is this line in the code. I was not sure what it does
https://github.com/resin-io-modules/drivelist/blob/master/scripts/win32.bat#L135

`Mountpoints` is a list of strings and I introduced `MountpointObjects` which is a list of objects. But list function works only with simple types:
https://github.com/resin-io-modules/drivelist/blob/master/scripts/win32.bat#L37

So I used the `Mountpoints` for condition check, but store actual data inside `MountpointObjects`.

Not the best solution, imho, but was a minimal change in the code

File: scripts/darwin.sh (hunk 1 above)
Comment (lurch) [inline scripts/darwin.sh]: Is that _guaranteed_ though, or is it just common convention? Perhaps there's a more reliable way of getting the volumeName?

**Comment by lurch** [comment]
Assigning this to @jviotti since (unfortunately for him) I think he knows more VBScript than any of the other Etcher team-members.

File: lib/parse.js (hunk 4 above)
Comment (lurch) [inline lib/parse.js]: Why does it filter out the line? Don't you want the volumeName to also make it into the JSON objects?

File: scripts/darwin.sh (hunk 1 above)
Comment (vohtaski) [inline scripts/darwin.sh]: Looks like not `guaranteed`. I have the following drives
```
volumeName: "/"
volumeName: "/Volumes/My Passport"
```

File: lib/parse.js (hunk 4 above)
Comment (vohtaski) [inline lib/parse.js]: yes, that's exactly what I want :)
But this line `/^(\s\s-\s)?[a-z]+:/g.test(line);` removes volumeName, since it's has uppercase and has 4 spaces before

**Comment by vohtaski** [comment]
I've added a better way of extracting volume name on mac

File: scripts/darwin.sh [hunk 5]
```diff
@@ -60,8 +60,11 @@ for disk in $DISKS; do
   else
     echo "mountpoints:"
     echo "$mountpoints" | while read -r mountpoint ; do
+      volume="$(echo "$mount_output" | perl -n -e'm{^('"${disk}"'(s[0-9]+)?) on (.*) \(.*\)$} && print "$1\n"')"
```
Comment (lurch) [inline scripts/darwin.sh]: Could this be modified to not use `perl` ?
We wouldn't want to see https://github.com/resin-io/etcher/issues/1288 getting re-introduced ;-)

Comment (vohtaski) [inline scripts/darwin.sh]: I think it can use sed or awk, but there is perl used just some lines above :)

**Comment by jhermsmeier** [comment]
@lurch, I think versionist will only do anything if a label is applied, hence no status report from it yet

**Comment by hedss** [comment]
@lurch @jhermsmeier   VersionBot only currently operates on Resin respositories. This is because it is not a public Integration. Unfortunately, in this case, we're trying to merge from an outside repo (`vohtaski:master`) and VersionBot has no access to it. So, we're going to need to rethink our workflow here. We could use a repo merge branch. I think this needs some discussion in a process call.

File: lib/parse.js (hunk 4 above)
Comment (lurch) [inline lib/parse.js]: Ahhh, I think this line is used to filter out any line that _doesn't_ look like YAML, in order to try and prevent e.g. any error messages included in the script-output from breaking the YAML parser.
But according to both http://www.yaml.org/spec/1.2/spec.html#id2759963 and http://www.yamllint.com/ the example in `README.md` of:
```
device: /dev/disk1
description: "Macintosh HD"
size: 249779191808
mountpoints:
  - path: /
raw: /dev/rdisk1
protected: False
system: True
```
would be 'more correct' if formatted as:
```
device: /dev/disk1
description: "Macintosh HD"
size: 249779191808
mountpoints:
  -
    path: /
raw: /dev/rdisk1
protected: False
system: True
```
which I guess means (if the scripts were all updated to output this 'more correct' format) the regex could be updated to `/^((\s{4})?[a-z]+:)|(\s{2}-$)/`. And I guess you could change `volumeName` to something all-lowercase like `label` ;-)

Although having thought about it some more, I'm not sure if adding `label` to `mountpoints` is entirely what you want, as it'll only give you volume-labels for partitions that are actually mounted, and volume-labels for unmounted partitions won't be shown. So I think the discussion in this PR actually overlaps with that in #98

File: scripts/darwin.sh (hunk 5 above)
Comment (lurch) [inline scripts/darwin.sh]: Ah, good point! I'd forgotten that #168 only removes one usage of `perl`, not all usages of perl ;-)

**Comment by oriolpuig** [comment]
Hi @vohtaski @lurch  and @jviotti,

Have you some update about it? 

I'm combining **drivelist** with **linux-mountutils** to mount and unmount in linux via NodeJS.

**Drivelist** reported me one device in `/dev/sdc` but If I do a  mount in bash, Linux says me `/dev/sdc1`. So, when I try to mount `/dev/sdc`, a **linux-mountutils** error displays "this device is already mounted or busy". 

That error is because drivelist doesn't provide the volumeName. 

Thanks for all your job.

BR

Oriol

**Comment by vohtaski** [comment]
@oriolpuig 
I am using this fork for our needs, I don't know if there is a plan to merge it

**Comment by lurch** [comment]
@oriolpuig `/dev/sdc` is the device name (i.e. the whole hard-drive) but `/dev/sdc1` is the partition name (i.e. the first (and maybe only?) partition on the drive). It's possible for a drive to contain multiple partitions. Etcher (which is what this module was written for) only needs the drive name (because it overwrites the whole drive at a low level). `linux-mountutils` (and the `mount` command) need the partition name, because they need to know which partition on the drive (there could be several) you actually want to mount.

**Comment by oriolpuig** [comment]
yes, @lurch and that's why I would like to have the `volumeName`, to concat the `deviceName` and volumeName and mount it with the other library.

**Comment by lurch** [comment]
@oriolpuig "volume name" is something else entirely, and I'm not sure it's actually what you're after. AFAIK volume name is effectively a text-label that you can add to a partition, _if_ the filesystem used on a partition supports such a feature.

**Comment by lurch** [comment]
I guess this PR is now woefully out of date, now that the Windows part of `drivelist` is implemented as a native C++ module.

**Comment by jhermsmeier** [comment]
Yeah, closing this, as it's not applicable anymore – yet we might want to look at adding that, if feasible.
